syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs

```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
-            chunkPlain    = Atom.StrPlain <$> takeWhile1 (not . isBodyChar)
-            chunkQuote    = bodyQuotes =<< takeWhile1 isQuote
-            bodyQuotes qs = Atom.StrPlain qs <$ failIf (Source.length qs == quoteLen)
-            chunk         = spanned $ choice [chunkPlain, chunkQuote] -- lineBreak
-            ending        = void $ takeWhile1 isQuote
-            end           = ending <|> (TokenStream.add =<< invalid (pure $ Invalid.Literal $ Invalid.String Invalid.NoClosingMark))
-            -- body          = Ast.Str . concat <$> flexBlock (many1' chunk)
-            chunks        = (<>) <$> many1' chunk <*> (concat <$> flexBlock (many1' chunk))
-            body          = Ast.Str <$> chunks
-        if quoteLen == 2 then pure (Ast.Str mempty) else body <* end
+        let chunkBody    = takeWhile1 isBodyChar
+            chunkQuote   = bodyQuotes =<< takeWhile1 isQuote
+            bodyQuotes a = a <$ failIf (quoteLen == Source.length a)
+            chunkPlain   = spanned $ Atom.StrPlain . mconcat <$> many1' (choice [chunkBody, chunkQuote])
+
+            quotes       = void $ takeWhile1 isQuote
+            end          = quotes <|> noEnd
+            chunks       = many1' chunkPlain <|> (mergeStrChunks <$> strBlock chunkPlain)
+            noEnd        = (TokenStream.add =<< invalid (pure $ Invalid.Literal $ Invalid.String Invalid.NoClosingMark))
+            body         = Ast.Str <$> chunks
+        if quoteLen == 2 then pure (Ast.Str mempty) else (body <* end) <|> (Ast.Str mempty <$ noEnd)
 
 
 --------------------
 -- === Layout === --
 --------------------
 
-flexBlock1 :: Lexer a -> Lexer (NonEmpty a)
+strBlock1 :: forall t. Lexer (Spanned (Atom.StrChunk t)) -> Lexer [Spanned (Atom.StrChunk t)]
+strBlock1 = \p -> let
+    line :: Bool -> Lexer (Spanned Delta, [Spanned (Atom.StrChunk t)])
+    line isFirst = do
+        off <- spanned $ newline *> Indent.indented
+        (off,) <$> many p
+
+    lines = do
+        ls <- (:) <$> line True <*> many (line False)
+        let sinds  = fst <$> ls
+            conts  = snd <$> ls
+            spans  = view Ast.span    <$> sinds
+            inds   = Ast.unsafeUnspan <$> sinds
+            (is' : inds')  = subtract minInd  <$> inds
+            minInd = unsafeMinimum inds
+            astLs  = (Atom.StrPlain . convert .            flip replicate ' ' $   is')
+                   : (Atom.StrPlain . convert . ('\n' :) . flip replicate ' ' <$> inds')
+            astLs' = zipWith Ast.Spanned spans astLs
+            conts' = zipWith (:) astLs' conts
+        pure $ concat conts'
+    in lines
+{-# INLINE strBlock1 #-}
+
+strBlock :: Lexer (Spanned (Atom.StrChunk t)) -> Lexer [Spanned (Atom.StrChunk t)]
+strBlock = option mempty . strBlock1
+{-# INLINE strBlock #-}
+
+
+mergeStrChunks :: [Spanned (Atom.StrChunk t)] -> [Spanned (Atom.StrChunk t)]
+mergeStrChunks = \case
+    []     -> []
+    (a:as) -> let
+        span   = view Ast.span a
+        spans  = view Ast.span <$> as
```

Comment:
There is a code repetition, consider refactor.

Reply:
This is just a lens accessor, it's not a code repetition.